esphome/components/uponor_smatrix/uponor_smatrix.cpp

```diff
@@ -0,0 +1,227 @@
+    return false;
+  }
+
+  ESP_LOGV(TAG, "Received packet: sys=%04X, dev=%04X, data=%s, crc=%04X", system_address, device_address,
+           format_hex(&packet[4], packet_len - 6).c_str(), crc);
+
+  // Detect or check system address
+  if (this->address_ == 0) {
+    ESP_LOGI(TAG, "Using detected system address 0x%04X", system_address);
+    this->address_ = system_address;
+  } else if (this->address_ != system_address) {
+    // This should never happen except if the system address was set or detected incorrectly, so warn the user.
+    ESP_LOGW(TAG, "Received packet from unknown system address 0x%04X", system_address);
+    return true;
+  }
+
+  // Handle packet
+  size_t data_len = (packet_len - 6) / 3;
+  if (data_len == 0) {
+    if (packet[4] == UPONOR_ID_REQUEST)
+      ESP_LOGVV(TAG, "Ignoring request packet for device 0x%04X", device_address);
+    return true;
+  }
+
+  // Decode packet payload data for easy access
+  UponorSmatrixData data[data_len];
+  for (int i = 0; i < data_len; i++) {
+    data[i].id = packet[(i * 3) + 4];
+    data[i].value = encode_uint16(packet[(i * 3) + 5], packet[(i * 3) + 6]);
+  }
+
+#ifdef USE_TIME
+  // Detect device that acts as time master if not set explicitely
+  if (this->time_device_address_ == 0 && data_len >= 2) {
+    // The first thermostat paired to the controller will act as the time master. Time can only be manually adjusted at
+    // this first thermostat. To synchronize time, we need to know its address, so we search for packets coming from a
+    // thermostat sending both room temperature and time information.
+    bool found_temperature = false;
+    bool found_time = false;
+    for (int i = 0; i < data_len; i++) {
+      if (data[i].id == UPONOR_ID_ROOM_TEMP)
+        found_temperature = true;
+      if (data[i].id == UPONOR_ID_DATETIME1)
+        found_time = true;
+      if (found_temperature && found_time) {
+        ESP_LOGI(TAG, "Using detected time device address 0x%04X", device_address);
+        this->time_device_address_ = device_address;
+        break;
+      }
+    }
+  }
+#endif
+
+  // Forward data to device components
+  bool found = false;
+  for (auto *device : this->devices_) {
+    if (device->address_ == device_address) {
+      found = true;
+      device->on_device_data(data, data_len);
+    }
+  }
+
+  // Log unknown device addresses
+  if (!found && !this->unknown_devices_.count(device_address)) {
+    ESP_LOGI(TAG, "Received packet for unknown device address 0x%04X ", device_address);
+    this->unknown_devices_.insert(device_address);
+  }
+
+  // Return true to reset buffer
+  return true;
+}
+
+bool UponorSmatrixComponent::send(uint16_t device_address, const UponorSmatrixData *data, size_t data_len) {
+  if (this->address_ == 0 || device_address == 0 || data == nullptr || data_len == 0)
+    return false;
+
+  // Assemble packet for send queue. All fields are big-endian except for the little-endian checksum.
+  std::vector<uint8_t> packet;
```

Comment:
We can reserve the underlying data array since we know how many bytes will be pushed
```suggestion
  std::vector<uint8_t> packet(6 + 3 * data_len);
```